### DISCUSSION TITLE
Mutations tested unreliably:  where to search for the bug?

### DISCUSSION THREAD

**Discussion by kevinmatthes** [sourcefrog/cargo-mutants#280] [Q&A] [answered]
I experienced on several different machines with two different code bases that the generated mutations are not always tested reliably.  Sometimes, mutations are caught by the same test suite, sometimes not, causing the CI to fail apparently for no reason.

Here is an example:

- [here](https://github.com/kevinmatthes/aeruginous-tbr/actions/runs/7933159115/job/21661287155), 9 of 9 mutants are caught; and
- [here](https://github.com/kevinmatthes/aeruginous-tbr/actions/runs/7936926802/job/21673152960), 1 of 9 mutants is not caught by the very same test suite.

The above example originates from two subsequent PRs which both do not have changed the test suite.  Thus, my question now is:  have I called ``cargo-mutants`` [the wrong way](https://github.com/kevinmatthes/aeruginous-tbr/actions/runs/7936926802/workflow#L97) or might this be an edge-case bug?

**Comment by kevinmatthes** [comment]
[Running the failed workflow again](https://github.com/kevinmatthes/aeruginous-tbr/actions/runs/7936926802/job/21674749736) seemed to have caught the mutant although I did not have changed the test suite either.  To really catch it once and for all, I will need to create another test case.

**Comment by sourcefrog** [comment]
I don't think you're calling it the wrong way, although you could try adding `--no-shuffle` to remove one source of variability. (Although arguably it's a feature that shuffling may highlight that there is non-deterministic behavior.)

Also I'd suggest changing your CI to upload `mutants.out` after the run, like in <https://mutants.rs/ci.html>, which will help enormously in debugging this kind of thing.

I'd also add `-vV` to the CI invocation at least temporarily to make it clearer what's happening. (Although this info is also present in the logs.)

What I think is most likely is that one of the tests is not quite deterministic. 

If you look in `mutants.out/logs` at the file for the mutant that is only sometimes caught that should show you which tests are failing sometimes.

**Comment by sourcefrog** [comment]
OK I could easily reproduce the nondeterminicity here, so it's not something about CI:

```
Found 9 mutants to test
ok       Unmutated baseline in 2.2s build + 0.1s test
 INFO Auto-set test timeout to 20s
caught   src/tar_archive.rs:58:9: replace TarArchive::create -> Result<()> with Ok(()) in 0.3s build + 0.1s test
caught   src/tar_archive.rs:89:9: replace TarArchive::extract -> Result<()> with Ok(()) in 0.3s build + 0.1s test
caught   src/tar_archive.rs:47:9: replace TarArchive::add_files -> Result<()> with Ok(()) in 0.3s build + 0.1s test
caught   src/tar_archive.rs:77:9: replace TarArchive::exists -> bool with false in 0.3s build + 0.1s test
caught   src/tar_archive.rs:101:9: replace TarArchive::list -> Result<Vec<PathBuf>> with Ok(vec![]) in 0.3s build + 0.1s test
caught   src/tar_archive.rs:125:9: replace TarArchive::remove -> Result<()> with Ok(()) in 0.3s build + 0.1s test
caught   src/tar_archive.rs:132:9: replace TarArchive::update -> Result<()> with Ok(()) in 0.3s build + 0.1s test
caught   src/tar_archive.rs:77:9: replace TarArchive::exists -> bool with true in 0.3s build + 0.1s test
caught   src/tar_archive.rs:101:9: replace TarArchive::list -> Result<Vec<PathBuf>> with Ok(vec![Default::default()]) in 0.3s build + 0.1s test
9 mutants tested in 5s: 9 caught
bobcat ~/src/mutants (main) 07:24
; c r mutants -d ~/src/aeruginous-tbr/ -vV
    Finished dev [unoptimized + debuginfo] target(s) in 0.05s
     Running `target/debug/cargo-mutants mutants -d /home/mbp/src/aeruginous-tbr/ -vV`
Found 9 mutants to test
ok       Unmutated baseline in 2.3s build + 0.1s test
 INFO Auto-set test timeout to 20s
caught   src/tar_archive.rs:77:9: replace TarArchive::exists -> bool with true in 0.3s build + 0.1s test
MISSED   src/tar_archive.rs:125:9: replace TarArchive::remove -> Result<()> with Ok(()) in 0.3s build + 0.1s test
caught   src/tar_archive.rs:89:9: replace TarArchive::extract -> Result<()> with Ok(()) in 0.3s build + 0.1s test
caught   src/tar_archive.rs:101:9: replace TarArchive::list -> Result<Vec<PathBuf>> with Ok(vec![Default::default()]) in 0.3s build + 0.1s test
caught   src/tar_archive.rs:47:9: replace TarArchive::add_files -> Result<()> with Ok(()) in 0.3s build + 0.1s test
caught   src/tar_archive.rs:132:9: replace TarArchive::update -> Result<()> with Ok(()) in 0.3s build + 0.1s test
caught   src/tar_archive.rs:101:9: replace TarArchive::list -> Result<Vec<PathBuf>> with Ok(vec![]) in 0.3s build + 0.1s test
caught   src/tar_archive.rs:58:9: replace TarArchive::create -> Result<()> with Ok(()) in 0.3s build + 0.1s test
caught   src/tar_archive.rs:77:9: replace TarArchive::exists -> bool with false in 0.3s build + 0.1s test
9 mutants tested in 5s: 1 missed, 8 caught
```

I think I can see in your tests a place where they are not hermetic.

Remember that Rust's test suite (regardless of cargo-mutants) runs multiple tests in parallel. So anything that affects global state of the process or the machine can cause interference between tests. 

In this particular case, interestingly, the non-hermicity never seems to cause a problem in normal runs, only under cargo-mutants, and presumably also if you had a naturally-occurring bug.

**Reply by kevinmatthes** [reply]
Thank you for your help!  I configured the asset upload as you suggested and found some other weird side effects of my code / test suite:  no matter which coverage tool I call, the results are always either random or incomplete; cargo-llvm-cov randomly does not execute certain (instructions of) tests and cargo-tarpaulin categorically refuses to check the macro-defined tests.  I start believing that something is horribly wrong with my code...  Would you please tell me where you found the not hermetic section in my tests?

Just for completeness:  https://github.com/kevinmatthes/aeruginous-tbr/actions/runs/7979616779 is the latest run which again could not catch mutants deterministically.  The strange aspect is that there actually is a test to catch a certain mutant but this test is not always executed -- neither by cargo-mutants nor by the two coverage tools from above.  What did I do wrong with the test design?

**Reply by sourcefrog** [reply]
The line that I thought was suspicious in https://github.com/kevinmatthes/aeruginous-tbr/blob/1df902d55c5eb4dbfe818756baa8e2454a4ca5bb/tests/tar_archive.rs#L28C1-L54C2

    let tar = TarArchive::new("tar_archive!life_cycle.tar"); 

This seems to be creating a tar file object with one fixed filename, unless there's something I missed that puts the file in a different directory per test. This will be in the source root directory.

If the test runs to completion it will create and then delete this file. If a test fails partway through then the file may be left on disk. If a later test runs in the same directory then it may fail because the file already exists, etc.

cargo-mutants will run the tests repeatedly with different mutations in the same directory. 

It looks like this crate has only a couple of tests but if you added more, and they create or edit files in the current directory then the possible interactions are even greater.

The simple fix is to use a temp file or a temp directory, e.g. through the rust tempfile crate.

I'm not sure off hand what could be causing the problems with cargo-llvm-cov, and I'm not sure what you mean by the macro-defined tests...

**Reply by sourcefrog** [reply]
>The strange aspect is that there actually is a test to catch a certain mutant but this test is not always executed -- neither by cargo-mutants nor by the two coverage tools from above. What did I do wrong with the test design?

Which mutant and which test are you talking about?

As far as I can see there are three tests and they are run for every viable mutant, based on the files in the `log/` directory in the `mutants.out` zip on that run.

```
running 3 tests
test exists_failure ... ok
test remove_failure ... ok
test life_cycle ... FAILED
```

**Reply by kevinmatthes** [reply]
> and I'm not sure what you mean by the macro-defined tests...

Short answer:  https://github.com/kevinmatthes/aeruginous-tbr/blob/734b2dbc8a13237abda6e86c8f557eb5d0586e6c/tests/application.rs#L204-L219.

Long answer:  there has been some progress to the repository; there are now 35 tests in total as well as an executable which has four different modes, triggered by various keywords each.  To simplify the testing of the keyword mapping to the associated application modes, I wrote a macro (see the link above) which expands to the actual set of tests.  These tests do not seem to be executed by the named coverage tools.  In another code base, https://github.com/kevinmatthes/aeruginous-rs, I apply even more macros in the tests and there never was any problem -- with none of the coverage tools.  Thus, there seems to be something special about that particular repository as multiple tools seem to have difficulties to work with the defined tests.

**Reply by kevinmatthes** [reply]
> The simple fix is to use a temp file or a temp directory, e.g. through the rust tempfile crate.

I will give this a try, thank you for the suggestion!

**Reply by kevinmatthes** [reply]
> Which mutant and which test are you talking about?

Test:  https://github.com/kevinmatthes/aeruginous-tbr/blob/734b2dbc8a13237abda6e86c8f557eb5d0586e6c/tests/application.rs#L90-L96

Mutant:  src/application.rs:100:9: replace Logic::tar_archive -> Result<()> with Ok(())

The mutant should be caught if there is a test which actually causes the ``tar_archive`` method to fail.  The test attempts to delete a TAR archive which does not exist, thus the ``tar_archive`` method exits with some value other than ``Ok(())`` as the underlying code does so, as well, in case that the archive to be removed does not exist.  This mutant is not always caught.

**Reply by sourcefrog** [reply, marked as answer]
What I would suggest there is:

Make that change by hand in your working tree; see if the tests catch it. 

I ran that here from the `feature/tar-handling-cli` branch and it says all the mutants were missed:

```
; cargo mutants -f application.rs --regex Logic --in-place
Found 5 mutants to test
ok       Unmutated baseline in 5.2s build + 0.1s test
 INFO Auto-set test timeout to 20s
MISSED   src/application.rs:87:9: replace Logic::resolve_files -> Result<()> with Ok(()) in 0.6s build + 0.1s test
MISSED   src/application.rs:76:31: replace == with != in Logic::main in 0.5s build + 0.1s test
MISSED   src/application.rs:66:9: replace Logic::main -> Result<()> with Ok(()) in 0.6s build + 0.1s test
MISSED   src/application.rs:100:9: replace Logic::tar_archive -> Result<()> with Ok(()) in 0.5s build + 0.1s test
```

I can also see the tests created a `application!application!create!tar_archive.tar` file which is an example of the non-hermiticity I talked about.

I ran `cargo t --test application` and I can see that it runs some tests and then

```
test application_mode::display::update ... ok
test application_mode::from_str::compress ... ok
test application_mode::from_str::content ... ok
test application_mode::from_str::create ... ok
test application_mode::from_str::decompress ... ok
Interact with Brotli, TAR, TAR.BR, and TBR archives

Usage: tbr [OPTIONS] <MODE> <ARCHIVE> [FILES]...

Arguments:
  <MODE>      The operation to perform on the considered archive
  <ARCHIVE>   The archive to interact with
  [FILES]...  The file(s) to add to the considered archive

Options:
  -d, --destination <DESTINATION>  The directory to unpack the archive's files into; defaulting to the current working directory, if omitted
  -h, --help                       Print help
```

I think you're calling a Clap function that causes the process to `exit(0)`. If you call this from the tests then, at whatever moment this is hit, it will look like the whole test target was successful.

**Reply by kevinmatthes** [reply]
Thank you, this solved the problem!
